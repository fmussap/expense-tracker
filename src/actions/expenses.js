import * as actions from './actions-type'
import uuid from 'uuid'

export const addExpense = (
  {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
  } = {}
) => {
  return ({
    type: actions.ADD_EXPENSE,
    expense: {
      id: uuid(),
      description,
      note,
      amount,
      createdAt
    }
  })
}

export const removeExpense = ({ id } = {}) => ({
  type: actions.REMOVE_EXPENSE,
  id
})

export const editExpense = (id, updates) => ({
  type: actions.EDIT_EXPENSE,
  id,
  updates
})
