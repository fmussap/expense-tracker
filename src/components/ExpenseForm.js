import React, { Component } from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

class ExpenseForm extends Component {
  constructor (props) {
    super()
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? props.expense.amount : 0,
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({ description }))
  }

  onAmountChange = (e) => {
    let amount = e.target.value
    if ( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      if (!amount) {
        amount = 0
      }
      this.setState(() => ({ amount }))
    }
  }

  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({ note }))
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }

  handleSubmit = (e) => {
    e.preventDefault()
    let error = ''
    if (!this.state.description) {
      error = 'Description field is required'
      this.setState(() => ({ error }))
    } else if (!this.state.amount) {
      error = 'Amount field is required'
      this.setState(() => ({ error }))
    } else {
      this.props.onSubmit({
        description: this.state.description,
        amount: this.state.amount,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf()
      })
    }
    
  }

  render () {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.handleSubmit}>
          <input
            type='text'
            placeholder='Description'
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type='number'
            placeholder='Amount'
            value={this.state.amount}
            min="0" step="0.01"
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder='Add a note for your expense'
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button type='submit'>{this.props.expense ? 'Edit Expense' : 'Add Expense'}</button>
        </form>
      </div>
    )
  }
}

export default ExpenseForm
