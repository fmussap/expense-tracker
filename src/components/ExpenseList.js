import React from 'react'
import { connect } from 'react-redux'

import ExpenseListItem from './ExpenseListItem'
import SelectExpenses from '../selectors/expenses-selector'

export const ExpenseList = (props) => {
  const renderList = () => {
    return (
      <div>
        <div className='list-header'>
          <div className='show-for-mobile'>Expenses</div>
          <div className='show-for-desktop'>Expense</div>
          <div className='show-for-desktop'>Amount</div>
        </div>
        <div className='list-body'>
          {props.expenses.length === 0
            ? (
              <div className='list-item list-item--message'>
                <span>no expenses</span>
              </div>
            )
            : props.expenses.map((expense) => {
              return <ExpenseListItem key={expense.id} {...expense} />
            })
          }
        </div>
      </div>
    )
  }
  return (
    <div className='content-container'>
      {renderList()}
    </div>
  )
}

const mapStateToProps = ({ expenses, filters }) => ({
  expenses: SelectExpenses(expenses, filters)
})

export default connect(mapStateToProps, undefined)(ExpenseList)
