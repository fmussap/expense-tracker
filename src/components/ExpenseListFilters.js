import React, { Component } from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters'

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  }

  handleInput = (e) => (
    this.props.setTextFilter(e.target.value)
  )

  handleSelect = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate()
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount()
    }
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  }

  render () {
    return (
      <div>
        <input type='text' value={this.props.filters.text} onChange={this.handleInput} />
        <select value={this.props.filters.sortBy} onChange={this.handleSelect}>
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          startDateId='start'
          endDateId='end'
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    )
  }
}

const mapStateToProps = ({ filters }) => ({
  filters
})

export default connect(mapStateToProps, {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
})(ExpenseListFilters)
