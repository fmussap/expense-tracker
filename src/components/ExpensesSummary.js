import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import { Link } from 'react-router-dom'

import SelectExpenses from '../selectors/expenses-selector'
import ExpensesTotal from '../selectors/expenses-total'

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  return (
    <div className='page-header'>
      <div className='content-container'>
        <h1 className='page-header__title'>
          Viewing <span>{expenseCount} </span>
          totalling <span>{expensesTotal}</span>
        </h1>
        <div className='page-header__actions'>
          <Link className='button' to='/create'>
            Add expense
          </Link>
        </div>
      </div>
    </div>
  )
}

const mapStateToProps = ({ expenses, filters }) => {
  const visibleExpenses = SelectExpenses(expenses, filters)
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: numeral(ExpensesTotal(visibleExpenses)).format('$0,0.00')
  }
}

export default connect(mapStateToProps, undefined)(ExpensesSummary)
