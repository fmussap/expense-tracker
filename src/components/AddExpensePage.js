import React from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'

export const AddExpensePage = (props) => {
  const handleSubmit = (expense) => {
    props.addExpense(expense)
    props.history.push('/')
  }
  return (
    <div>
      <h1>Add Expense</h1>
      <ExpenseForm onSubmit={handleSubmit} />
    </div>
  )
}

export default connect(undefined, {
  addExpense
})(AddExpensePage)
