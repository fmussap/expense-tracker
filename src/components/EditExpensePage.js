import React from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses'

export const EditExpensePage = (props) => {
  const handleSubmit = (expense) => {
    props.editExpense(props.expense.id, expense)
    props.history.push('/')
  }
  const handleClick = (e) => {
    props.removeExpense({ id: props.expense.id })
    props.history.push('/')
  }
  return (
    <div>
      <h1>Edit Expense</h1>
      <ExpenseForm onSubmit={handleSubmit} expense={props.expense} />
      <button onClick={handleClick}>Remove</button>
    </div>
  )
}

const mapStateToProps = ({ expenses }, props) => ({
  expense: expenses.find((expense) => expense.id === props.match.params.id)
})

export default connect(mapStateToProps, {
  editExpense,
  removeExpense
})(EditExpensePage)
