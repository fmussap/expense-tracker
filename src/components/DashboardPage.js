import React from 'react'
import { Link } from 'react-router-dom'

import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'

const DashboardPage = () => (
  <div>
    <Link to='/create'>
      create
    </Link>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)

export default DashboardPage
