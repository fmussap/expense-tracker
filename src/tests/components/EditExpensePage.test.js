import React from 'react'
import { shallow } from 'enzyme'

import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'

let spy, history, wrapper

beforeEach(() => {
  spy = jest.fn()
  history = {
    push: jest.fn()
  }
  wrapper = shallow(<EditExpensePage
    editExpense={spy}
    removeExpense={spy}
    history={history}
    expense={expenses[0]}
  />)
})

test('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle handleSubmit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0])
  expect(spy).toHaveBeenLastCalledWith(expenses[0].id, expenses[0])
  expect(history.push).toHaveBeenLastCalledWith('/')
})

test('should handle handleClick', () => {
  wrapper.find('button').simulate('click')
  expect(spy).toHaveBeenLastCalledWith({ id: expenses[0].id })
  expect(history.push).toHaveBeenLastCalledWith('/')
})