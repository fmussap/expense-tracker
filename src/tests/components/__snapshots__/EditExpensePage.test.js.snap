// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 5,
        "createdAt": 1514282568444,
        "description": "banana",
        "id": 1,
        "note": "note test",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
